checker/src/org/checkerframework/checker/optional/OptionalVisitor.java

```diff
@@ -0,0 +1,340 @@
+    private boolean isOptionalCreation(MethodInvocationTree methInvok) {
+        return OptionalUtils.isMethodInvocation(methInvok, "of", 1, atypeFactory)
+                || OptionalUtils.isMethodInvocation(methInvok, "ofNullable", 1, atypeFactory);
+    }
+
+    /**
+     * @return true iff expression is a call to Optional elimination: get, orElse, orElseGet,
+     *     orElseThrow.
+     */
+    private boolean isOptionalElimation(MethodInvocationTree methInvok) {
+        return OptionalUtils.isMethodInvocation(methInvok, "get", 0, atypeFactory)
+                || OptionalUtils.isMethodInvocation(methInvok, "orElse", 1, atypeFactory)
+                || OptionalUtils.isMethodInvocation(methInvok, "orElseGet", 1, atypeFactory)
+                || OptionalUtils.isMethodInvocation(methInvok, "orElseThrow", 1, atypeFactory);
+    }
+
+    @Override
+    public Void visitConditionalExpression(ConditionalExpressionTree node, Void p) {
+        handleTernaryIsPresentGet(node);
+        return super.visitConditionalExpression(node, p);
+    }
+
+    /**
+     * Part of rule #3.
+     *
+     * <p>Pattern match for: {@code VAR.isPresent() ? VAR.get().METHOD() : VALUE}
+     *
+     * <p>Prefer: {@code VAR.map(METHOD).orElse(VALUE);}
+     */
+    public void handleTernaryIsPresentGet(ConditionalExpressionTree node) {
+
+        ExpressionTree condExpr = TreeUtils.skipParens(node.getCondition());
+        ExpressionTree trueExpr = TreeUtils.skipParens(node.getTrueExpression());
+        ExpressionTree falseExpr = TreeUtils.skipParens(node.getFalseExpression());
+
+        if (!isCallToIsPresent(condExpr)) {
+            return;
+        }
+        ExpressionTree receiver = TreeUtils.getReceiverTree(condExpr);
+
+        if (trueExpr.getKind() != Kind.METHOD_INVOCATION) {
+            return;
+        }
+        ExpressionTree trueReceiver = TreeUtils.getReceiverTree(trueExpr);
+        if (!isCallToGet(trueReceiver)) {
+            return;
+        }
+        ExpressionTree getReceiver = TreeUtils.getReceiverTree(trueReceiver);
+
+        // What is a better way to do this than string comparison?
+        if (receiver.toString().equals(getReceiver.toString())) {
+            ExecutableElement ele = TreeUtils.elementFromUse((MethodInvocationTree) trueExpr);
+
+            checker.report(
+                    Result.warning(
+                            "prefer.map.and.orelse",
+                            receiver,
+                            // The literal "CONTAININGCLASS::" is gross.
+                            // TODO: add this to the error message.
+                            //ElementUtils.getQualifiedClassName(ele);
+                            ele.getSimpleName(),
+                            falseExpr),
+                    node);
+        }
+    }
+
+    /** Return true if the two trees represent the same expression. */
+    private boolean sameExpression(ExpressionTree tree1, ExpressionTree tree2) {
```

Comment:
This method is currently not called. Should it be used above in `if (receiver.toString().equals(getReceiver.toString()))`?

Reply:
Done.